README.md
[![CRAN status](https://www.r-pkg.org/badges/version/spots)](https://CRAN.R-project.org/package=spots) [![Build status](https://ci.appveyor.com/api/projects/status/lbdrt72vyiqwxxff/branch/main?svg=true)](https://ci.appveyor.com/project/stevexniu/spots/branch/main) [![Build Status](https://app.travis-ci.com/stevexniu/spots.svg?branch=main)](https://app.travis-ci.com/stevexniu/spots) [![codecov](https://codecov.io/gh/stevexniu/spots/branch/main/graph/badge.svg?token=7KF4D3GGUB)](https://codecov.io/gh/stevexniu/spots)

# spots: Spatial Component Analysis <img src="man/figures/logo.png" align="right" width="150"/>

The **```spots```** package is designed for spatial omics (10x Visium, etc.) data analysis. 

It performs various statistical analyses and tests, including spatial component analysis (SCA), both global and local spatial statistics, such as univariate and bivariate Moran's I, Getis-Ord Gi* statistics, etc.

See <a href="https://doi.org/10.1101/2022.03.15.484516" target="_blank">Integrated protein and transcriptome high-throughput spatial profiling (2022)</a> for more details.

Installation
-----
Install from CRAN release:

``` r
install.packages("spots")
```

Install from Github:

``` r
install.packages("devtools")
devtools::install_github("stevexniu/spots")
```

Install the light-weight ```spots-feather``` without eign decomposition (```SCA``` functionality).

``` r
install.packages("devtools")
devtools::install_github("stevexniu/spots", ref = "feather")
```

Usage
-----
See tutorials:

[Get Started](https://stevexniu.github.io/spots/articles/get_started.html)

[Spatial Statistics](https://stevexniu.github.io/spots/articles/spatial_statistics.html)

© [X. Steve Niu](https://github.com/stevexniu) from [Landau Lab](https://www.landaulab.org), Weill Cornell Medicine and New York Genome Center


NEWS.md
# spots-feather v0.1.0 (2022-05-10)
+ Add light-weight ```spots-feather``` branch.

# spots v0.1.0 (2022-04-27)

+ Initial release of ```spots```.


tests/testthat.R
library(testthat)
library(spots)

test_check("spots")


tests/testthat/test-local.R
# Test for local Moran's I
library(stats)
library(ape)
set.seed(42)
test.X <- matrix(runif(300), 30, 10)
test.W <- matrix(runif(900), 30, 30)
test.r <- apply(test.X, 2, ape::Moran.I, weight = test.W)
test.moransi <- unlist(lapply(test.r, "[[", "observed"))
test.mean <- unlist(lapply(test.r, "[[", "expected"))
test.sd <- unlist(lapply(test.r, "[[", "sd"))
test.pval <- unlist(lapply(test.r, "[[", "p.value"))

test_that("Local Moran's I calculated correctly", {
  test.locali<- LocalMoransI(test.X, test.W)
  expect_equal(colSums(test.locali$Local.Morans.I), test.moransi)
})

test_that("Permutation local Moran's I calculated correctly", {
  test.permi <- apply(test.X, 2, PermutationLocalI, W = test.W)
  expect_equal(unlist(lapply(test.permi, function(x) sum(x[['Local.Morans.I']]))), test.moransi)
})


tests/testthat/test-stats.R
# Test for statistics calculation
library(stats)
set.seed(42)
test.x <- runif(1000)
test.y <- runif(1000)
test.q <- matrix(seq(0.05, 0.95, 0.05))
test.norm <- qnorm(test.q)

test_that("Permutation correlation calculated correctly", {
  test.stats <- PermutationCorr(test.x, test.y)
  test.r <- cor.test(test.x, test.y)
  expect_equal(test.stats$correlation, test.r$estimate, ignore_attr = TRUE)
  expect_equal(test.stats$p.val > 0.05, test.r$p.value > 0.05)
})

test_that("Z score p-values calculated correctly", {
  test.stats <- ZPvalue(Z = test.norm, alternative = "less")
  expect_equal(test.stats$p.val, test.q)
  test.stats <- ZPvalue(Z = test.norm, alternative = "greater")
  expect_equal(test.stats$p.val, 1-test.q)
  test.stats <- ZPvalue(Z = test.norm, alternative = "two.sided", p.adjust.method = "bonferroni")
  expect_equal(test.stats$p.val, ifelse(test.q > 0.5, 2*(1-test.q), 2*test.q))
  expect_equal(test.stats$p.adj, p.adjust(test.stats$p.val, method = "bonferroni"), ignore_attr = TRUE)
})


R/hnn.R
#' @include stats.R
#'
NULL

#' Hexagonal nearest neighbor
#'
#' Calculate hexagonal nearest neighbors.
#'
#' @param dist.hnn A hexagonal nearest neighbor distance matrix.
#' @param k Number of neighbors.
#' @param include.self Whether to include self as 1st neighbor, default is \code{TRUE}.
#' @return A list containing the following:
#' \itemize{
#'   \item knn.idx, an n x k matrix for the nearest neighbor indice.
#'   \item knn.dist, an n x k matrix for the nearest neighbor hexagonal distances.
#'   \item dist.mat, a connectivity-based distance matrix.
#' }
#' @export
#' @concept hnn
#' @examples {
#' data.use <- quakes[1:100,]
#' dist.use <- as.matrix(dist(data.use[,1:2]))
#' res <- HnnNeighbor(dist.use, k = 10)
#' }
#' @references
#' Middleton, L. & Sivaswamy, J.
#' Edge detection in a hexagonal-image processing framework. Image Vis. Comput. 19, 1071–1081 (2001)
#'
HnnNeighbor <- function(dist.hnn, k, include.self = TRUE){
  n <- nrow(x = dist.hnn)
  k.start <- ifelse(test = include.self, yes = 1, no = 2)
  k.end <- ifelse(test = include.self, yes = k, no = k+1)
  knn.idx <- matrix(data = 0, ncol = k, nrow = n)
  knn.dist <- matrix(data = 0, ncol = k, nrow = n)
  dist.mat <- dist.hnn
  for(i in 1:n){
    knn.idx[i,] <- order(dist.hnn[i,])[k.start:k.end]
    knn.dist[i,] <- dist.hnn[i,knn.idx[i,]]
    dist.mat[i,-knn.idx[i,]] <- Inf
  }
  rownames(x = knn.idx) <- rownames(x = dist.hnn)
  return(list(knn.idx = knn.idx, knn.dist = knn.dist, dist.mat = dist.mat))
}

#' Hexagonal nearest neighbor index
#'
#' Get hexagonal nearest neighbor indices.
#'
#' @param dist.hnn A hexagonal nearest neighbor distance matrix.
#' @param k Number of neighbors.
#' @param include.self Whether to include self as 1st neighbor, default is \code{TRUE}.
#' @return An n x k matrix for the nearest neighbor indice.
#' @keywords internal
#'
HnnNeighborIndex <- function(dist.hnn, k, include.self = TRUE){
  k.start <- ifelse(test = include.self, yes = 1, no = 2)
  k.end <- ifelse(test = include.self, yes = k, no = k+1)
  knn.idx <- t(x = apply(X = dist.hnn, MARGIN = 1, FUN = function(x){
    order(x)[k.start:k.end]
  }))
  rownames(x = knn.idx) <- rownames(x = dist.hnn)
  return(knn.idx)
}

#' Hexagonal nearest neighbor weight
#'
#' Calculate hexagonal nearest neighbor weights using Gaussian filter.
#'
#' @param dist.hnn A hexagonal nearest neighbor distance matrix.
#' @param dist.k The maximum distance used to calculate the weight. Default is \code{NULL} and all neighbor weights are calculated.
#' @param mu The mean of Gaussian filter, default is 0.
#' @param sigma The standard deviation of Gaussian filter, default is 1.
#' @return A weight matrix.
#' @importFrom stats dnorm
#' @export
#' @concept hnn
#' @examples {
#' data.use <- quakes[1:100,]
#' dist.use <- as.matrix(dist(data.use[,1:2]))
#' res <- HnnWeight(dist.use, mu = 0, sigma = 0.5)
#' }
#'
HnnWeight <- function(dist.hnn, dist.k = NULL, mu = 0, sigma=1){
  if(!is.null(x = dist.k)) dist.hnn[dist.hnn>dist.k] <- Inf
  hnn.weight <- dnorm(x = dist.hnn, mean = mu, sd = sigma)
  return(hnn.weight)
}

#' Hexagonal nearest neighbor based imputation
#'
#' Data imputation and smoothing using hexagonal nearest neighbor.
#'
#' @param data A data matrix with features as rows and observations as columns.
#' @param dist.hnn A hexagonal nearest neighbor distance matrix.
#' @param dist.k The maximum distance used to calculate the weight. Default is \code{NULL} and all neighbor weights are calculated.
#' @param mu The mean of Gaussian filter, default is 0.
#' @param sigma The standard deviation of Gaussian filter, default is 1.
#' @return Imputed data.
#' @seealso \code{\link[spots]{HnnWeight}}
#' @export
#' @concept hnn
#' @examples {
#' data.use <- quakes[1:100,]
#' dist.use <- as.matrix(dist(data.use[,1:2]))
#' # transpose the data to have features in rows and observations in columns
#' res <- HnnImpute(t(data.use[,3:4]), dist.use)
#' }
#'
HnnImpute <- function(data, dist.hnn, dist.k = NULL, mu = 0, sigma=1){
  if(!is.null(x = dist.k)) dist.hnn[dist.hnn>dist.k] <- Inf
  hnn.weight <- HnnWeight(dist.hnn = dist.hnn, dist.k = dist.k, mu = mu, sigma = sigma)
  hnn.weight <- sweep(x = hnn.weight, MARGIN = 2, STATS = colSums(hnn.weight), FUN = "/")
  data <- data %*% hnn.weight
  return(data)
}

#' 10x Genomics Visium hexagonal nearest neighbor
#'
#' Get hexagonal nearest neighbor distance between 10x Visium spatial barcodes.
#'
#' @param path Path to the directory contains or to download the HNN distance for Visium barcodes, i.e. "~/Downloads".
#' @param barcodes 10x Visium whitelisted spatial barcodes.
#' @return A hexagonal distance matrix for selected barcodes.
#' @seealso \code{\link[spots]{available.data}}  \code{\link[spots]{LoadData}}.
#' @export
#' @concept hnn
#' @examples \dontrun{
#' barcodes <- c("AAACAACGAATAGTTC-1", "AAACTTAATTGCACGC-1")
#' VisiumHnn("~/Downloads/", barcodes)
#' }
#'
VisiumHnn <- function(path, barcodes){
  Visium.hnn.dist <- LoadData(path, "Visium.HNN")
  dist.use <- as.matrix(Visium.hnn.dist[barcodes, barcodes])
  return(dist.use + t(dist.use))
}


R/sca.R
#' @import Matrix
#'
NULL

#' Bivariate Moran's I using Wartenberg's M matrix
#'
#' Calculate bivariate Moran's I using Wartenberg's method.
#' @param X A matrix with observations as rows and features as columns.
#' @param W A weight matrix across all observations, i.e inverse of a pairwise distance matrix.
#' @return A spatial cross-correlation matrix.
#' @export
#' @concept sca
#' @examples {
#' data.use <- quakes[1:100,]
#' W <- 1/as.matrix(dist(data.use[,1:2]))
#' diag(W) <- 0
#' M <- WartenbergM(data.use[,3:4], W)
#' }
#' @references
#' Wartenberg, D. Multivariate spatial correlation:
#' A method for exploratory geographical analysis. Geogr. Anal. 17, 263–283 (1985)
#'
WartenbergM <- function(X, W){
  N <- nrow(X)
  X <- scale(X, center = TRUE, scale = FALSE)
  W.sum <- sum(W)
  W.sum[W.sum==0] <- 1
  W <- W/W.sum
  S <- sqrt(colSums(X*X)/N)
  M <- t(X)%*%W%*%X/tcrossprod(S)
  if(is(M,"denseMatrix")) M <- as.matrix(M)
  return(M)
}

#' Bivariate Moran's I using Lee's L measurement
#'
#' Calculate bivariate Moran's I using Lee's L measurement.
#'
#' @param X A matrix with observations as rows and features as columns.
#' @param W A weight matrix across all observations, i.e inverse of a pairwise distance matrix.
#' @return A spatial cross-correlation matrix.
#' @export
#' @concept sca
#' @examples {
#' data.use <- quakes[1:100,]
#' W <- 1/as.matrix(dist(data.use[,1:2]))
#' diag(W) <- 0
#' L <- LeeL(data.use[,3:4], W)
#' }
#' @references
#' Lee, S.-I. Developing a bivariate spatial association measure:
#' An integration of Pearson’s r and Moran's I. J. Geogr. Syst. 3, 369–385 (2001)
#'
LeeL <- function(X, W){
  N <- nrow(X)
  X <- scale(X, center = TRUE, scale = TRUE)
  W.rowsum <- rowSums(W)
  W.rowsum[W.rowsum==0] <- 1
  W <- W/W.rowsum
  L <- t(X)%*%t(W)%*%W%*%X/N
  if(is(L,"denseMatrix")) L <- as.matrix(L)
  return(L)
}

#' Spatial cross-correlation
#'
#' Calculate spatial cross-correlation matrix.
#'
#' @param X A matrix with observations as rows and features as columns.
#' @param W A weight matrix across all observations, i.e inverse of a pairwise distance matrix.
#' @param method Method used to calculate spatial cross-correlation.
#' \itemize{
#'   \item M, using the Wartenburg's M (Default).
#'   \item L, using the Lee's L.
#' }
#' @return A spatial cross-correlation matrix.
#' @export
#' @concept sca
#' @seealso \code{\link[spots]{WartenbergM}} \code{\link[spots]{LeeL}}.
#' @examples {
#' data.use <- quakes[1:100,]
#' W <- 1/as.matrix(dist(data.use[,1:2]))
#' diag(W) <- 0
#' M <- SpatialXCorr(data.use[,3:4], W, method = "M")
#' L <- SpatialXCorr(data.use[,3:4], W, method = "L")
#' }
#' @references
#' Wartenberg, D. Multivariate spatial correlation:
#' A method for exploratory geographical analysis. Geogr. Anal. 17, 263–283 (1985)
#'
#' Lee, S.-I. Developing a bivariate spatial association measure:
#' An integration of Pearson’s r and Moran's I. J. Geogr. Syst. 3, 369–385 (2001)
#'
SpatialXCorr <- function(X, W, method = c("M", "L")){
  xcor <- switch(match.arg(method),
                 M = WartenbergM,
                 L = LeeL
                 )
  xcor(X, W)
}


R/data.R
#' @importFrom tools md5sum
#' @importFrom utils download.file
#'
NULL

#' Available data
#'
#' Available data can be loaded using \code{\link[spots]{LoadData}} function.
#'
#' @format A data frame with 2 rows (datasets) and 5 columns (attributes):
#' \describe{
#'   \item{filename}{Dataset}
#'   \item{md5}{MD5 checksum}
#'   \item{url}{URL link}
#'   \item{date}{Date updated}
#'   \item{size}{Data size}
#' }
#' @concept data
#'
"available.data"

#' Load selected data
#'
#' If not previously run, it will download the selected data.
#'
#' @param file.path Path to the directory contains or to download the data object, i.e. "~/Downloads".
#' @param data.use Selected data to load. See \code{\link[spots]{available.data}}.
#' \itemize{
#'   \item Visium.HNN, hexagonal nearest neighbor distance matrix of
#'   \href{https://www.10xgenomics.com/spatial-transcriptomics}{10x Visium} 4992 whitelist spatial barcodes.
#'   \item Visium.Brain, sample \href{https://www.10xgenomics.com/resources/datasets/mouse-brain-serial-section-1-sagittal-posterior-1-standard-1-0-0}{10x Visium Sagittal-Posterior Mouse Brain}
#'   data with 3353 spatial barcodes and 2000 highly variable genes, load as a \href{https://satijalab.org/seurat/articles/spatial_vignette.html}{Seurat object}.
#' }
#' @return Selected data:
#' @export
#' @concept data
#' @examples \dontrun{
#' Visium.hnn.dist <- LoadData("~/Downloads", "Visium.HNN")
#' }
#'
LoadData <- function(file.path, data.use = c("Visium.HNN", "Visium.Brain")){
  meta.data <- spots::available.data[match.arg(data.use, choices = rownames(spots::available.data)), ]
  file.path <- paste0(file.path, "/", meta.data[['filename']])
  file.md5 <- md5sum(file.path)
  if(is.na(file.md5) | file.md5 != meta.data[['md5']]){
    message("Downloading ", data.use)
    download.file(url = meta.data[['url']], destfile = file.path)
  }
  data.use <- readRDS(file.path)
  return(data.use)
}




R/moransi.R
#' @include stats.R sca.R
#'
NULL

#' Univariate Moran's I
#'
#' Calculate univariate (canonical) Moran's I.
#'
#' @param X A matrix with observations as rows and features as columns.
#' @param W A weight matrix across all observations, i.e inverse of a pairwise distance matrix.
#' @param normalize Whether to normalize the weight matrix such that each row adds up to one. Default is \code{TRUE}.
#' @param alternative Alternative hypothesis used, default is \code{two.sided}.
#' @param p.adjust.method Method used for multiple comparisons correction, default is \code{BH}. See \code{\link[stats]{p.adjust}}.
#' @return A list containing the following:
#' \itemize{
#'   \item Morans.I, the Moran's I.
#'   \item Z.I, the Z score of Moran's I.
#'   \item X, data matrix used for calculating Moran's I.
#'   \item Y, a matrix of spatial lags.
#'   \item Expected.I, the expectation of Moran's I under the null hypothesis.
#'   \item SD.I, the standard deviation of Moran's I under the null hypothesis.
#'   \item p.val, p-values.
#'   \item p.adj, adjusted p-values.
#'   \item normalize, whether to normalize the weight matrix.
#'   \item alternative, alternative hypothesis used.
#'   \item p.adjust.method, method used for multiple comparisons correction.
#' }
#' @export
#' @concept moransi
#' @examples {
#' data.use <- quakes[1:100,]
#' W <- 1/as.matrix(dist(data.use[,1:2]))
#' diag(W) <- 0
#' res <- UnivariateMoransI(data.use[,3:4], W)
#' }
#' @references
#' Moran, P. A. P. Notes on continuous stochastic phenomena. Biometrika 37, 17–23 (1950)
#'
UnivariateMoransI <- function(X, W, normalize = TRUE, alternative = c("two.sided", "less", "greater"), p.adjust.method = "BH"){
  alternative <- match.arg(arg = alternative)
  X <- scale(X, center = TRUE, scale = FALSE)
  N <- nrow(X)
  if(normalize){
    W.rowsum <- rowSums(W)
    W.rowsum[W.rowsum==0] <- 1
    W <- W/W.rowsum
    W2 <- N^2
    S2 <- sum((1+colSums(W))^2)
  } else {
    W.sum <- sum(W)
    W <- N/W.sum*W
    W2 <- W.sum^2
    S2 <- sum((rowSums(W)+colSums(W))^2)
  }
  Y <- W%*%X
  if(is(Y,"denseMatrix")) Y <- as.matrix(Y)
  X2 <- colSums(X*X)
  I <- colSums(X*Y)/X2
  EI <- -1/(N-1)
  S1 <- 0.5*sum((W+t(W))^2)
  S3 <- (colSums(X^4)*N)/(X2^2)
  S4 <- (N^2-3*N+3)*S1-N*S2+3*W2
  S5 <- N*(N-1)*S1-2*N*S2+6*W2
  SDI <- sqrt((N*S4-S3*S5) / ((N-1)*(N-2)*(N-3)*W2) - EI^2)
  Z <- (I-EI)/SDI
  P.list <- ZPvalue(as.matrix(Z), alternative, p.adjust.method)
  return(list(Morans.I = I, Z.I = Z, X = X, Y = Y, Expected.I = EI, SD.I = SDI,  p.val = P.list[[1]][,1], p.adj = P.list[[2]][,1],
             normalize = normalize,alternative = alternative, p.adjust.method = p.adjust.method))
}

#' Permutation based Moran's I
#'
#' Calculate permutation based p-value for Moran's I.
#'
#' @param x A numerical vector.
#' @param W A weight matrix across all observations, i.e inverse of a pairwise distance matrix.
#' @param n The number of permutations to be conducted, set to 999 by default.
#' @param seed Random seed used. Default is 1.
#' @param normalize Whether to normalize the weight matrix such that each row adds up to one. Default is \code{TRUE}.
#' @param return.permutation Return permutations. Default is \code{FALSE}.
#' @return A list containing the following:
#' \itemize{
#'   \item Morans.I, the Moran's I.
#'   \item p.val, permutation based p-value.
#'   \item return.permutation, permutation used if returned.
#' }
#' @export
#' @concept moransi
#' @importFrom stats sd
#' @examples {
#' data.use <- quakes[1:100,]
#' W <- 1/as.matrix(dist(data.use[,1:2]))
#' diag(W) <- 0
#' res <- PermutationMoransI(data.use[,3], W)
#' }
#'
PermutationMoransI <- function(x, W, n = 999, seed = 1, normalize = TRUE, return.permutation = FALSE){
  x <- x - mean(x)
  N <- length(x)
  if(normalize){
    W.rowsum <- rowSums(W)
    W.rowsum[W.rowsum==0] <- 1
    W <- W/W.rowsum
  } else {
    W.sum <- sum(W)
    W.sum[W.sum==0] <- 1
    W <- N/W.sum*W
  }
  y <- (W%*%x)[,1]
  sd.x <- sd(x)
  sd.y <- sd(y)
  r <- PermutationCorr(x, y, n, seed, return.permutation)
  if(return.permutation) return.permutation = r[[3]]*sd.y/sd.x
  return(list(Morans.I = r[[1]]*sd.y/sd.x, p.val = r[[2]], return.permutation = return.permutation))
}

#' Bivariate Moran's I
#'
#' Calculate bivariate (multivariate) Moran's I using Wartenberg's method.
#'
#' @param X A matrix with observations as rows and features as columns.
#' @param W A weight matrix across all observations, i.e inverse of a pairwise distance matrix.
#' @param alternative Alternative hypothesis used, default is \code{two.sided}.
#' @param p.adjust.method Method used for multiple comparisons correction, default is \code{BH}. See \code{\link[stats]{p.adjust}}.
#' @return A list containing the following:
#' \itemize{
#'   \item Morans.I, the Moran's I.
#'   \item Z.I, the Z score of Moran's I.
#'   \item Expected.I, the expectation of Moran's I under the null hypothesis.
#'   \item SD.I, the standard deviation of Moran's I under the null hypothesis.
#'   \item alternative, alternative hypothesis used.
#'   \item p.adjust.method, method used for multiple comparisons correction.
#' }
#' @export
#' @concept moransi
#' @examples {
#' data.use <- quakes[1:100,]
#' W <- 1/as.matrix(dist(data.use[,1:2]))
#' diag(W) <- 0
#' res <- BivariateMoransI(data.use[,3:4], W)
#' }
#' @references
#' Wartenberg, D. Multivariate spatial correlation:
#' A method for exploratory geographical analysis. Geogr. Anal. 17, 263–283 (1985)
#'
#' Czaplewski, R. L. Expected Value and Variance of Moran’s Bivariate Spatial Autocorrelation Statistic for a Permutation Test.
#' (U.S. Department of Agriculture, Forest Service, Rocky Mountain Forest and Range Experiment Station, 1993).
#'
BivariateMoransI <- function(X, W, alternative = c("two.sided", "less", "greater"), p.adjust.method = "BH"){
  alternative <- match.arg(arg = alternative)
  I <- WartenbergM(X, W)
  I.stats <- BivariateMoransIStats(X, W)
  Z <- (I-I.stats[[1]])/I.stats[[2]]
  P.list <- ZPvalue(Z, alternative, p.adjust.method)
  return(list(Morans.I = I, Z.I = Z, Expected.I = I.stats[[1]], SD.I = I.stats[[2]],
              p.val = P.list[[1]], p.adj = P.list[[2]], alternative = alternative, p.adjust.method = p.adjust.method))
}

#' Moran's I using ordinary least squares (OLS)
#'
#' Calculate Moran's I as linear regression using ordinary least squares (OLS).
#'
#' @param X A matrix with observations as rows and features as columns.
#' @param W A weight matrix across all observations, i.e inverse of a pairwise distance matrix.
#' @param normalize Whether to normalize the weight matrix such that each row adds up to one. Default is \code{TRUE}.
#' @param alternative Alternative hypothesis used, default is \code{two.sided}.
#' @param p.adjust.method Method used for multiple comparisons correction, default is \code{BH}. See \code{\link[stats]{p.adjust}}.
#' @return A list containing the following:
#' \itemize{
#'   \item Morans.I, the Moran's I.
#'   \item Z.I, the Z score of Moran's I.
#'   \item X, data matrix used for calculating Moran's I.
#'   \item Y, a matrix of spatial lags.
#'   \item Expected.I, the expectation of Moran's I under the null hypothesis.
#'   \item SD.I, the standard deviation of Moran's I under the null hypothesis.
#'   \item p.val, p-values.
#'   \item p.adj, adjusted p-values.
#'   \item alternative, alternative hypothesis used.
#'   \item p.adjust.method, method used for multiple comparisons correction.
#' }
#' @export
#' @concept moransi
#' @examples {
#' data.use <- quakes[1:100,]
#' W <- 1/as.matrix(dist(data.use[,1:2]))
#' diag(W) <- 0
#' res <- OLSMoransI(data.use[,3:4], W)
#' }
#' @references
#' Anselin, L. Local indicators of spatial association-LISA. Geogr. Anal. 27, 93–115 (2010)
#'
OLSMoransI <- function(X, W, normalize = TRUE, alternative = c("two.sided", "less", "greater"), p.adjust.method = "BH"){
  alternative <- match.arg(arg = alternative)
  X <- scale(X, center = TRUE, scale = FALSE)
  N <- nrow(X)
  if(normalize){
    W.rowsum <- rowSums(W)
    W.rowsum[W.rowsum==0] <- 1
    W <- W/W.rowsum
  } else {
    W.sum <- sum(W)
    W.sum[W.sum==0] <- 1
    W <- N/W.sum*W
  }
  Y <- W%*%X
  if(is(Y,"denseMatrix")) Y <- as.matrix(Y)
  XY <- crossprod(X,Y)
  X2 <- colSums(X*X)
  I <- XY/X2
  S <- sqrt(tcrossprod(1/X2,X2))
  I.stats <- BivariateMoransIStats(X,W)
  I.stats[[1]] <- I.stats[[1]]*S
  I.stats[[2]] <- I.stats[[2]]*S
  Z <- (I-I.stats[[1]])/I.stats[[2]]
  P.list <- ZPvalue(Z, alternative, p.adjust.method)
  return(list(Morans.I = I, Z.I = Z, X = X, Y = Y, Expected.I = I.stats[[1]], SD.I = I.stats[[2]],
              p.val = P.list[[1]], p.adj = P.list[[2]], alternative = alternative, p.adjust.method = p.adjust.method))
}

#' Fitted values of Moran's I
#'
#' Calculate fitted values of Moran's I as linear regression.
#'
#' @param X A matrix with observations as rows and features as columns.
#' @param Y A matrix of spatial lags.
#' @param I A matrix of Moran's I, output by \code{\link[spots]{OLSMoransI}}..
#' @param all Whether to use multivariate Moran's I, default is \code{FALSE}.
#' @return A list containing the following:
#' \itemize{
#'   \item residuals, residuals of fitted Moran's I.
#'   \item fitted.values, fitted values of Moran's I.
#'   \item all.fitted.values, a three-dimensional tensor of all fitted multivariate Moran's I, if calculated.
#'   \item all.residuals, a three-dimensional tensor of all residuals using fitted multivariate Moran's I, if calculated.
#' }
#' @export
#' @concept moransi
#' @examples {
#' data.use <- quakes[1:100,]
#' W <- 1/as.matrix(dist(data.use[,1:2]))
#' diag(W) <- 0
#' res <- OLSMoransI(data.use[,3:4], W)
#' res.fitted <- FitMoransI(res$X, res$Y, res$Morans.I)
#' }
#' @references
#' Anselin, L. Local indicators of spatial association-LISA. Geogr. Anal. 27, 93–115 (2010)
#'
FitMoransI<- function(X, Y, I, all = FALSE){
  N <- nrow(X)
  M <- ncol(X)
  names.dim <- dimnames(X)
  if(all){
    All.X <- array(rep(X, M), dim = c(N, M, M))
    All.Y.hat <- array(0, dim = c(N, M, M), dimnames = list(names.dim[[1]], names.dim[[2]], paste0(names.dim[[2]],".y")))
    All.R <- array(0, dim = c(N, M, M), dimnames = list(names.dim[[1]], names.dim[[2]], paste0(names.dim[[2]],".y")))
    Y.hat <- matrix(0, nrow = N, ncol = M, dimnames = names.dim)
    R <- matrix(0, nrow = N, ncol = M, dimnames = names.dim)
    for(i in 1:M){
      All.Y.hat[,,i] <- t(t(All.X[,,i])*I[,i])
      All.R[,,i] <- Y[,i]-All.Y.hat[,,i]
      Y.hat[,i] <- All.Y.hat[,i,i]
      R[,i] <- All.R[,i,i]
    }
    list(residuals = R, fitted.values = Y.hat, all.fitted.values = All.Y.hat, all.residuals = All.R)
  } else {
    Y.hat <- t(t(X)*diag(I))
    R <- Y-Y.hat
    list(residuals = R, fitted.values = Y.hat)
  }
}

#' Process irregular data shape (between two groups) for Moran's I
#'
#' Process irregular data shape for Moran's I between two groups.
#'
#' @param X A matrix with observations as rows and features as columns.
#' @param W A weight matrix across all observations, i.e inverse of a pairwise distance matrix.
#' @param group1 The indices or names for the first group of observations.
#' @param group2 The indices or names for the second group of observations.
#' @return A list containing the following:
#' \itemize{
#'   \item X, new data matrix of group1 and group2.
#'   \item W, new weight matrix of group1 and group2.
#' }
#' @export
#' @concept moransi
#' @examples {
#' data.use <- quakes[1:100,]
#' W <- 1/as.matrix(dist(data.use[,1:2]))
#' diag(W) <- 0
#' res <- IrregularData(data.use[,3:4], W, 1:10, 21:30)
#' }
#'
IrregularData <- function(X, W, group1, group2){
  X <- X[c(group1, group2),]
  X <- X[,colSums(X)!=0]
  W[group1, group1] <- 0
  W[group2, group2] <- 0
  W <- W[c(group1, group2), c(group1, group2)]
  return(list(X = X, W = W))
}

#' Moran's I between two groups (irregular data shape)
#'
#' Calculate Moran's I between two groups with irregular data shape.
#'
#' @param X A matrix with observations as rows and features as columns.
#' @param W A weight matrix across all observations, i.e inverse of a pairwise distance matrix.
#' @param group1 The indices or names for the first group of observations.
#' @param group2 The indices or names for the second group of observations.
#' @param OLS Whether to use \code{\link[spots]{OLSMoransI}}, default is \code{FALSE} and \code{\link[spots]{BivariateMoransI}} is used.
#' @param normalize Whether to normalize the weight matrix such that each row adds up to one. Default is \code{TRUE}.
#' @param alternative Alternative hypothesis used, default is \code{two.sided}.
#' @param p.adjust.method Method used for multiple comparisons correction, default is \code{BH}. See \code{\link[stats]{p.adjust}}.
#' @return A list containing the output from \code{\link[spots]{BivariateMoransI}} or \code{\link[spots]{OLSMoransI}}.
#' @export
#' @concept moransi
#' @examples {
#' data.use <- quakes[1:100,]
#' W <- 1/as.matrix(dist(data.use[,1:2]))
#' diag(W) <- 0
#' res <- IrregularMoransI(data.use[,3:4], W, 1:10, 21:30)
#' }
#'
IrregularMoransI <- function(X, W, group1, group2, OLS = FALSE, normalize = TRUE, alternative = c("two.sided", "less", "greater"), p.adjust.method = "BH"){
  alternative <- match.arg(alternative)
  new.data <- IrregularData(X, W, group1, group2)
  if(OLS){
    OLSMoransI(new.data[[1]], new.data[[2]], normalize, alternative, p.adjust.method)
  } else {
    BivariateMoransI(new.data[[1]], new.data[[2]], alternative, p.adjust.method)
  }
}
